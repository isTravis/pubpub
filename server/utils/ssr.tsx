import React from 'react';
import ReactDOMServer from 'react-dom/server';
import * as ReactBeautifulDnD from 'react-beautiful-dnd';
import { Collection, InitialData, Attribution } from 'types';
import { isProd } from 'utils/environment';

export const renderToNodeStream = (res, reactElement) => {
	res.setHeader('content-type', 'text/html');
	res.write('<!DOCTYPE html>');
	ReactBeautifulDnD.resetServerContext();
	return ReactDOMServer.renderToNodeStream(reactElement).pipe(res);
};

type MetaProps = {
	initialData: InitialData;
	title: string;
	contextTitle?: string | null;
	description?: string | null;
	image?: string | null;
	attributions?: Attribution[];
	doi?: string | null;
	publishedAt?: null | Date;
	unlisted?: boolean;
	collection?: Collection;
	pdfDownloadUrl?: string;
	textAbstract?: string;
	notes?: string[];
	canonicalUrl?: string;
};

const contributorRoles = ['Writing – Review & Editing', 'Editor', 'Series Editor'];
const sortAttributions = (foo: Attribution, bar: Attribution) => {
	if (foo.order < bar.order) {
		return -1;
	}
	if (foo.order > bar.order) {
		return 1;
	}
	if (foo.createdAt < bar.createdAt) {
		return 1;
	}
	if (foo.createdAt > bar.createdAt) {
		return -1;
	}
	return 0;
};

export const generateMetaComponents = (metaProps: MetaProps) => {
	const {
		initialData,
		title,
		contextTitle,
		description,
		image,
		attributions,
		doi,
		publishedAt,
		unlisted,
		collection,
		pdfDownloadUrl,
		textAbstract,
		notes,
		canonicalUrl,
	} = metaProps;

	const {
		title: communityTitle,
		citeAs: communityCiteAs,
		publishAs: communityPublisher,
		twitter: communityTwitter,
	} = initialData.communityData;

	const url = `https://${initialData.locationData.hostname}${initialData.locationData.path}`;
	const isPub = !!initialData.scopeData?.elements?.activePub;
	const useCollectionTitle = !isPub && collection?.title;
	const favicon = initialData.communityData.favicon;
	const imageOrAvatar = image || initialData.communityData.avatar;
	const twitterCardType = image ? 'summary_large_image' : 'summary';
	const titleWithContext = contextTitle ? `${title} · ${contextTitle}` : title;
	let outputComponents: any[] = [];
	if (!initialData.locationData.isBasePubPub) {
		outputComponents = [
			...outputComponents,
			<link
				key="rss1"
				rel="alternate"
				type="application/rss+xml"
				title={`${title} RSS Feed`}
				href={`https://${initialData.locationData.hostname}/rss.xml`}
			/>,
		];
	}

	if (title) {
		outputComponents = [
			...outputComponents,
			<title key="t1">{titleWithContext}</title>,
			<meta
				key="t2"
				property="og:title"
				content={useCollectionTitle ? collection.title ?? undefined : title}
			/>,
			<meta key="t3" name="twitter:title" content={titleWithContext} />,
			<meta key="t4" name="twitter:image:alt" content={titleWithContext} />,
			<meta
				key="t5"
				name="citation_title"
				content={useCollectionTitle ? collection.title ?? undefined : title}
			/>,
			<meta
				key="t6"
				name="dc.title"
				content={useCollectionTitle ? collection.title ?? undefined : title}
			/>,
		];
	}

	if (communityTitle) {
		outputComponents = [
			...outputComponents,
			<meta key="sn1" property="og:site_name" content={communityTitle} />,
		];
	}

	if (communityTitle && !collection) {
		outputComponents = [
			...outputComponents,
			<meta key="sn2" name="citation_journal_title" content={communityTitle} />,
		];
	}

	if (url) {
		outputComponents = [
			...outputComponents,
			<meta key="u1" property="og:url" content={url} />,
			<meta
				key="u2"
				property="og:type"
				content={collection?.kind === 'book' ? 'book' : isPub ? 'article' : 'website'}
			/>,
		];
	}

	if (collection) {
		outputComponents = [
			...outputComponents,
			<meta key="c0" name="citation_publisher" content={communityPublisher || 'PubPub'} />,
		];
		if (collection.kind === 'issue') {
			outputComponents = [
				...outputComponents,
				<meta
					key="c1"
					name="citation_journal_title"
					content={communityCiteAs ?? undefined}
				/>,
				<meta key="c2" name="citation_volume" content={collection.metadata?.volume} />,
				<meta key="c3" name="citation_issue" content={collection.metadata?.issue} />,
				<meta
					key="c4"
					name="citation_issn"
					content={collection.metadata?.electronicIssn}
				/>,
				<meta key="c5" name="citation_issn" content={collection.metadata?.printIssn} />,
				<meta
					key="c6"
					name="citation_date"
					content={collection.metadata?.publicationDate}
				/>,
			];
		}
		if (collection.kind === 'book') {
			outputComponents = [
				...outputComponents,
				<meta
					key="c7"
					name="citation_book_title"
					content={collection.title ?? undefined}
				/>,
				<meta key="c8" name="citation_isbn" content={collection.metadata?.isbn} />,
				<meta
					key="c9"
					name="citation_date"
					content={collection.metadata?.publicationDate}
				/>,
			];
		}
		if (collection.kind === 'conference') {
			outputComponents = [
				...outputComponents,
				<meta
					key="c10"
					name="citation_conference_title"
					content={collection.title ?? undefined}
				/>,
				<meta
					key="c11"
					name="citation_conferenceName"
					content={collection.title ?? undefined}
				/>,
				<meta
					key="c12"
					name="citation_date"
					content={collection.metadata?.date ?? undefined}
				/>,
			];
		}
	}

	if (pdfDownloadUrl) {
		outputComponents = [
			...outputComponents,
			<meta key="dl1" name="citation_pdf_url" content={pdfDownloadUrl} />,
		];
	}

	if (textAbstract) {
		outputComponents = [
			...outputComponents,
			<meta key="a1" name="citation_abstract" content={textAbstract} />,
		];
	}

	if (communityTwitter) {
		outputComponents = [
			...outputComponents,
			<meta name="twitter:creator" content={`@${communityTwitter}`} />,
		];
	}

	if (description) {
		outputComponents = [
			...outputComponents,
			<meta key="d1" name="description" content={description} />,
			<meta key="d2" property="og:description" content={description} />,
			<meta key="d3" name="twitter:description" content={description} />,
		];
	}

	if (imageOrAvatar) {
		outputComponents = [
			...outputComponents,
			<meta key="i1" property="og:image" content={imageOrAvatar} />,
			<meta key="i2" property="og:image:url" content={imageOrAvatar} />,
			<meta key="i3" property="og:image:width" content="500" />,
			<meta key="i4" name="twitter:image" content={imageOrAvatar} />,
		];
	}

	if (favicon) {
		outputComponents = [
			...outputComponents,
			<link key="f1" rel="icon" type="image/png" sizes="256x256" href={favicon} />,
		];
	}

	if (attributions) {
		const authors: Attribution[] = [];
		const contributors: Attribution[] = [];

		const getPrimaryRole = (contributor: Attribution) => contributor.roles?.[0];

		attributions.sort(sortAttributions).forEach((attribution) => {
			if (!attribution.isAuthor) {
				return;
			}
			const primaryRole = getPrimaryRole(attribution);
			if (primaryRole && contributorRoles.includes(primaryRole)) {
				contributors.push(attribution);
			} else {
				authors.push(attribution);
			}
		});

		const citationAuthorTags = authors.map((author) => {
			if (author.user) {
				return (
					<meta
						key={`author-cite-${author.id}`}
						name="citation_author"
						content={author.user.fullName}
					/>
				);
			}
			return (
				<meta
					key={`author-cite-${author.id}`}
					name="citation_author"
					content={author.name ?? undefined}
				/>
			);
		});
		const dcAuthorTags = authors.map((author) => {
			if (author.user) {
				return (
					<meta
						key={`author-dc-${author.id}`}
						name="dc.creator"
						content={author.user.fullName}
					/>
				);
			}
			return (
				<meta
					key={`author-dc-${author.id}`}
					name="dc.creator"
					content={author.name ?? undefined}
				/>
			);
		});
		const contributorRoleTags = contributors.map((contributor) => {
			if (contributor.user) {
				return (
					<meta
						key={`editor-cite-${contributor.id}`}
						name="citation_editor"
						content={contributor.user.fullName}
					/>
				);
			}
			return (
				<meta
					key={`editor-cite-${contributor.id}`}
					name="citation_editor"
					content={contributor.name ?? undefined}
				/>
			);
		});
		outputComponents = [
			...outputComponents,
			citationAuthorTags,
			dcAuthorTags,
			contributorRoleTags,
		];
	}

	if (publishedAt) {
		const googleScholarPublishedAt = `${publishedAt.getFullYear()}/${
			publishedAt.getMonth() + 1
		}/${publishedAt.getDate()}`;
		const dcPublishedAt = `${publishedAt.getFullYear()}-${publishedAt.getMonth()}-${publishedAt.getDate()}`;
		outputComponents = [
			...outputComponents,
			<meta key="pa1" property="article:published_time" content={String(publishedAt)} />,
			<meta key="pa2" property="dc.date" content={dcPublishedAt} />,
			<meta key="pa3" name="citation_publication_date" content={googleScholarPublishedAt} />,
			<meta key="pub1" property="dc.publisher" content={communityPublisher || 'PubPub'} />,
		];
	}

	const finalDoi = doi || collection?.metadata?.doi;

	if (finalDoi) {
		outputComponents = [
			...outputComponents,
			<meta key="doi1" name="citation_doi" content={`doi:${finalDoi}`} />,
			<meta key="doi2" property="dc.identifier" content={`doi:${finalDoi}`} />,
			<meta key="doi3" property="prism.doi" content={`doi:${finalDoi}`} />,
		];
	}
	if (notes) {
		const citationNoteTags = notes.map((note, i) => {
			// https://github.com/yannickcr/eslint-plugin-react/issues/1123
			// eslint-disable-next-line react/no-array-index-key
			return <meta key={`n${i}`} name="citation_reference" content={note} />;
		});
		outputComponents = [...outputComponents, citationNoteTags];
	}
	if (!isProd() || unlisted) {
		outputComponents = [
			...outputComponents,
			<meta key="un1" name="robots" content="noindex,nofollow" />,
		];
	}

	if (canonicalUrl) {
		outputComponents = [...outputComponents, <link rel="canonical" href={canonicalUrl} />];
	}

	outputComponents = [
		...outputComponents,
		<meta key="misc1" property="fb:app_id" content="924988584221879" />,
		<meta key="misc2" name="twitter:card" content={twitterCardType} />,
		<meta key="misc3" name="twitter:site" content="@pubpub" />,
	];

	return outputComponents;
};
