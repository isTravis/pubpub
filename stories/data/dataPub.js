export default {
	id: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
	slug: '6pewbpry',
	title: 'Defining the Dimensions of the “Space” of Computing',
	description:
		'The first computing machines were so large they filled entire rooms. Today they are ubiquitous, built invisibly into our environments. While it is tempting to view this change within a predetermined space of progress, we can still shape the future on our own terms.',
	avatar: 'https://assets.pubpub.org/9pi1c5c6/41554230047846.jpeg',
	useHeaderImage: true,
	firstPublishedAt: '2019-04-22T03:43:24.218Z',
	lastPublishedAt: '2019-04-22T03:43:24.218Z',
	draftEditHash: 'eyy8hltx',
	draftViewHash: 's5mjfpdz',
	doi: null,
	labels: null,
	isCommunityAdminManaged: true,
	review: null,
	downloads: null,
	communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
	createdAt: '2019-04-02T18:12:05.538Z',
	updatedAt: '2019-04-30T15:23:17.211Z',
	branches: [
		{
			createdAt: '2019-04-29T22:01:27.918Z',
			id: '93504929-a39b-4d97-a618-83e00dc666ca',
			shortId: 1,
			title: 'draft',
			description: null,
			submissionAlias: null,
			order: 0,
			communityAdminPermissions: 'none',
			publicPermissions: 'edit',
			permissions: [
				{
					id: '00ed4bc1-a408-4725-87f4-b07a41b2cfb5',
					permissions: 'edit',
					userId: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
					branchId: '93504929-a39b-4d97-a618-83e00dc666ca',
					pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
					createdAt: '2019-04-02T18:12:05.543Z',
					updatedAt: '2019-04-02T18:12:05.543Z',
					user: {
						id: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
						firstName: 'Travis',
						lastName: 'Cohen',
						fullName: 'Travis Cohen',
						avatar: null,
						slug: 'traviscohen',
						initials: 'TC',
						title: 'Travis',
					},
				},
			],
			isActive: true,
			viewHash: 'cat',
			discussHash: 'dog',
			editHash: 'fish',
		},
		{
			createdAt: '2019-04-29T22:01:27.919Z',
			id: 'ebd27c37-27ee-4c75-b907-d99263a8641a',
			shortId: 2,
			title: 'public',
			description: null,
			submissionAlias: 'Submit for Review',
			order: 1,
			communityAdminPermissions: 'none',
			publicPermissions: 'edit',
			permissions: [],
			isActive: false,
			viewHash: 'cat2',
			discussHash: 'dog2',
			editHash: 'fish2',
		},
	],
	managers: [
		{
			id: '00ed4bc1-a408-4725-87f4-b07a41b2cfb5',
			userId: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:05.543Z',
			updatedAt: '2019-04-02T18:12:05.543Z',
			user: {
				id: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
				firstName: 'Travis',
				lastName: 'Cohen',
				fullName: 'Travis Cohen',
				avatar: null,
				slug: 'traviscohen',
				initials: 'TC',
				title: 'Travis',
			},
		},
		{
			id: '213eb90a-c6da-4617-8178-369b0a019cf1',
			userId: 'fc4d0b74-0fb0-413e-aea4-a9aecb3c1154',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:44:03.927Z',
			updatedAt: '2019-04-02T18:44:03.927Z',
			user: {
				id: 'fc4d0b74-0fb0-413e-aea4-a9aecb3c1154',
				firstName: 'Amy',
				lastName: 'Graves',
				fullName: 'Amy Graves',
				avatar: 'https://assets.pubpub.org/5k1fj9pu/51539111048243.jpg',
				slug: 'amy-graves',
				initials: 'AG',
				title: 'copy editor',
			},
		},
	],
	attributions: [
		{
			id: '146c7a07-43a7-49e1-8b04-734bbc7166c9',
			name: 'Weiwei Hsu',
			avatar: null,
			title: null,
			order: 0.25,
			isAuthor: true,
			roles: ['Writing – Original Draft Preparation'],
			userId: null,
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:48.718Z',
			updatedAt: '2019-04-02T18:12:53.295Z',
			user: {
				id: '146c7a07-43a7-49e1-8b04-734bbc7166c9',
				initials: 'W',
				fullName: 'Weiwei Hsu',
				firstName: 'Weiwei',
				lastName: 'Hsu',
				avatar: null,
				title: null,
			},
		},
		{
			id: '4f309ba3-5337-49f6-adee-8c921ae50d1d',
			name: 'Hugh Dubberly',
			avatar: null,
			title: null,
			order: 0.5,
			isAuthor: true,
			roles: ['Writing – Original Draft Preparation'],
			userId: null,
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:43.574Z',
			updatedAt: '2019-04-02T18:12:56.267Z',
			user: {
				id: '4f309ba3-5337-49f6-adee-8c921ae50d1d',
				initials: 'H',
				fullName: 'Hugh Dubberly',
				firstName: 'Hugh',
				lastName: 'Dubberly',
				avatar: null,
				title: null,
			},
		},
	],
	collectionPubs: [
		{
			id: '88000923-6c19-401a-acf5-98effe5c9c5b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			collectionId: '65fc8561-31ea-4e2c-8067-b315dd7c933f',
			contextHint: null,
			rank: null,
			isPrimary: false,
			createdAt: '2019-04-24T20:43:15.375Z',
			updatedAt: '2019-04-24T20:43:15.375Z',
			collection: {
				id: '65fc8561-31ea-4e2c-8067-b315dd7c933f',
				title: '5. Essays in Exploration: Resisting Reduction',
				isRestricted: true,
				isPublic: true,
				pageId: 'f7c79073-8110-40c2-9d4f-9d090673cec0',
				communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
				metadata: null,
				kind: 'tag',
				doi: null,
				createdAt: '2019-01-22T03:16:13.644Z',
				updatedAt: '2019-01-22T03:17:58.495Z',
				page: {
					id: 'f7c79073-8110-40c2-9d4f-9d090673cec0',
					title: '5. Essays in Exploration: Resisting Reduction',
					slug: 'issue-5',
				},
			},
		},
		{
			id: '6cfea932-1701-4d4e-86b9-f85758dad57b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			collectionId: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
			contextHint: null,
			rank: null,
			isPrimary: false,
			createdAt: '2019-04-24T20:43:15.375Z',
			updatedAt: '2019-04-24T20:43:15.375Z',
			collection: {
				id: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
				title: 'Home',
				isRestricted: true,
				isPublic: true,
				pageId: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
				communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
				metadata: null,
				kind: 'tag',
				doi: null,
				createdAt: '2018-10-15T23:51:00.296Z',
				updatedAt: '2018-10-15T23:51:00.296Z',
				page: {
					id: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
					title: 'Home',
					slug: '',
				},
			},
		},
	],
	discussions: [],
	activeBranch: {
		createdAt: '2019-04-29T22:01:27.918Z',
		id: '93504929-a39b-4d97-a618-83e00dc666ca',
		shortId: 1,
		title: 'draft',
		description: null,
		submissionAlias: null,
		order: 0,
		communityAdminPermissions: 'none',
		publicPermissions: 'edit',
		permissions: [],
		isActive: true,
		viewHash: 'cat2',
		discussHash: 'dog2',
		editHash: 'fish2',
	},
	isManager: true,
	isEditor: true,
	isViewer: false,
	isBranchAdmin: false,
	isStaticDoc: false,
	citationData: {
		pub: {
			apa:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry</div></div>',
			harvard:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry</div></div>',
			vancouver:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">    <div class="csl-left-margin">1. </div><div class="csl-right-inline">Hsu W, Dubberly H. Defining the Dimensions of the “Space” of Computing. JoDS V6 Dev [Internet]. 2019 Apr 30; Available from: https://jods.mitpress.mit.edu/pub/6pewbpry</div>  </div></div>',
			bibtex:
				'<div class="csl-bib-body"><div class="csl-entry">@article{Hsu2019Defining,<ul style="list-style-type:none"><li>journal = {JoDS V6 Dev},</li><li>note = {https://jods.mitpress.mit.edu/pub/6pewbpry},</li><li>title = {Defining the Dimensions of the “Space” of Computing},</li><li>url = {https://jods.mitpress.mit.edu/pub/6pewbpry},</li><li>author = {Hsu, Weiwei and Dubberly, Hugh},</li><li>date = {2019-04-30},</li><li>year = {2019},</li><li>month = {4},</li><li>day = {30},</li></ul>}</div></div>',
		},
		version: {
			apa:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div></div>',
			harvard:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div></div>',
			vancouver:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">    <div class="csl-left-margin">1. </div><div class="csl-right-inline">Hsu W, Dubberly H. Defining the Dimensions of the “Space” of Computing. JoDS V6 Dev [Internet]. 2019 Apr 30; Available from: https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div>  </div></div>',
			bibtex:
				'<div class="csl-bib-body"><div class="csl-entry">@article{Hsu2019Defining,<ul style="list-style-type:none"><li>journal = {JoDS V6 Dev},</li><li>note = {https://jods.mitpress.mit.edu/pub/6pewbpry/draft},</li><li>title = {Defining the Dimensions of the “Space” of Computing},</li><li>url = {https://jods.mitpress.mit.edu/pub/6pewbpry/draft},</li><li>author = {Hsu, Weiwei and Dubberly, Hugh},</li><li>date = {2019-04-30},</li><li>year = {2019},</li><li>month = {4},</li><li>day = {30},</li></ul>}</div></div>',
		},
	},
	firebaseToken:
		'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJodHRwczovL2lkZW50aXR5dG9vbGtpdC5nb29nbGVhcGlzLmNvbS9nb29nbGUuaWRlbnRpdHkuaWRlbnRpdHl0b29sa2l0LnYxLklkZW50aXR5VG9vbGtpdCIsImlhdCI6MTU1NjYzODU3NCwiZXhwIjoxNTU2NjQyMTc0LCJpc3MiOiJmaXJlYmFzZS1hZG1pbnNkay1pZTd5ZUBwdWJwdWItdjYuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLCJzdWIiOiJmaXJlYmFzZS1hZG1pbnNkay1pZTd5ZUBwdWJwdWItdjYuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLCJ1aWQiOiJiMjQyZjYxNi03YWFhLTQ3OWMtOGVlNS0zOTMzZGNmNzA4NTkifQ.AFdYAfq3y1F1lgLfC0HKvfg4ZkqOP6draEEHBVDJM1ioxh8N9guJ6PM9rieoKry2UxNRCY3t9BttYXIffotOJLjj3L6MKq0FaOoEBycPZcrhqEQX43oTna54Zt54WD2ZKqGKO0TT1iUAdxJLJfqfthKe_o7IkXlc27z7uu09ek2RGBAvuiS10Jzc4j5urNQuLh6UFEVdoMaxBollkjEwK2wV333Df4TZ9jyfZCw7lR8VTPyXoXmVJbMUccLNM-i2tNO6FEqJ-C132qL19uGcw4LQtjwzSV0MQJJAhs16ZeEKPTkDF-XuOL-41ycq2tH0DBUlhHjAlhgbs53hCynVRg',
	mode: 'manage',
};
