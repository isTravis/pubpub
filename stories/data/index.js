export const accentDataDark = {
	accentColor: '#D13232',
	accentTextColor: '#FFF',
	accentActionColor: '#A72828',
	accentHoverColor: '#BC2D2D',
	accentMinimalColor: 'rgba(209, 50, 50, 0.15)',
};

export const accentDataLight = {
	accentColor: '#26E0D0',
	accentTextColor: '#000',
	accentActionColor: '#51E6D9',
	accentHoverColor: '#3BE3D4',
	accentMinimalColor: 'rgba(38, 224, 208, 0.15)',
};

import initialData from './dataInitial.js';
export const { locationData, loginData } = initialData;

export { default as communityData } from './dataCommunity.js';
export { default as collectionData } from './dataCollection.js';
export { default as pubData } from './dataPub.js';
export { plainDoc, imageDoc, fullDoc } from './dataDocs.js';
