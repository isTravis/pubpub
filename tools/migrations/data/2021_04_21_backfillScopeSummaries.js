import { Community, Collection, Pub } from 'server/models';
import { summarizeCollection, summarizeCommunity, summarizePub } from 'server/scopeSummary/queries';

import { forEachInstance } from '../util';

export const up = async () => {
	await forEachInstance(Pub, (pub) => summarizePub(pub.id, false), 25);
	await forEachInstance(Collection, (collection) => summarizeCollection(collection.id), 25);
	await forEachInstance(Community, (community) => summarizeCommunity(community.id, 25));
};
