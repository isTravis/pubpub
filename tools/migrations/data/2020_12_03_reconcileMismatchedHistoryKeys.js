/* eslint-disable no-console */
import { Op } from 'sequelize';

import { sequelize, Branch, Release } from 'server/models';
import { getBranchDoc } from 'server/utils/firebaseAdmin';
import { forEach } from '../util';

const bucketByProperty = (array, property) => {
	const res = {};
	array.forEach((el) => {
		const propValue = el[property];
		if (!res[propValue]) {
			res[propValue] = [];
		}
		res[propValue].push(el);
	});
	return res;
};

const handleReleasesForPub = async (pubId, releases) => {
	const allReleasesForPub = await Release.findAll({
		where: { pubId: pubId },
		order: [['createdAt', 'ASC']],
	});
	const branches = await Branch.findAll({ where: { pubId: pubId } });
	const draftBranch = branches.find((br) => br.title === 'draft');
	const publicBranch = branches.find((br) => br.title === 'public');
	console.log(`=== ${pubId} ===`);
	await Promise.all(
		releases.map(async (release) => {
			const { sourceBranchKey, historyKey } = release;
			if (
				typeof sourceBranchKey === 'number' &&
				typeof historyKey === 'number' &&
				sourceBranchKey !== historyKey
			) {
				const relatedMergeIndex = allReleasesForPub.findIndex(
					(otherRelease) => otherRelease.id === release.id,
				);

				const [
					{ doc: releaseDoc },
					{ doc: sourceBranchKeyDoc },
					{ doc: historyKeyDoc },
				] = await Promise.all([
					getBranchDoc(pubId, publicBranch.id, relatedMergeIndex),
					getBranchDoc(pubId, draftBranch.id, sourceBranchKey),
					getBranchDoc(pubId, draftBranch.id, historyKey),
				]);

				const releaseDocString = JSON.stringify(releaseDoc);
				const sourceBranchKeyMatches =
					releaseDocString === JSON.stringify(sourceBranchKeyDoc);
				const historyKeyMatches = releaseDocString === JSON.stringify(historyKeyDoc);
				if (historyKeyMatches) {
					console.log('History key is already correct');
				} else if (sourceBranchKeyMatches) {
					console.log('Setting historyKey = sourceBranchKey');
					await Release.update(
						{ historyKey: sourceBranchKey },
						{ where: { id: release.id } },
					);
				} else {
					console.log(
						'Neither historyKey nor sourceBranchKey points to the correct document',
					);
				}
			}
		}),
	);
};

module.exports = async () => {
	const releasesWithMismatchedKeys = await Release.findAll({
		where: {
			sourceBranchKey: {
				[Op.ne]: null,
			},
			historyKey: {
				[Op.and]: [{ [Op.ne]: -1 }, { [Op.ne]: sequelize.col('sourceBranchKey') }],
			},
		},
	});
	const releasesByPubId = bucketByProperty(releasesWithMismatchedKeys, 'pubId');
	await forEach(Object.entries(releasesByPubId), ([pubId, releases]) =>
		handleReleasesForPub(pubId, releases),
	);
};
