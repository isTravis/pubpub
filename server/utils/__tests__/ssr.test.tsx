import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import dataInitial from 'utils/storybook/data/dataInitial';
import { pubData, attributionsData } from 'utils/storybook/data';
import { generateMetaComponents } from '../ssr';

Enzyme.configure({ adapter: new Adapter() });
const defaultProps = {
	initialData: dataInitial,
	attributions: pubData.attributions,
	title: 'Test Pub',
};

const MetaComponents = (props) => <>{generateMetaComponents(props)}</>;

describe('generateMetaComponents', () => {
	it('generates citation_author metadata for every author WITHOUT a contributor role as their primary role', () => {
		const attributions = pubData.attributions.map((att) => ({ ...att }));
		attributions[0].roles = ['Editor'];
		attributions[1].roles = ['Test role', 'Editor'];

		const wrap = shallow(<MetaComponents {...defaultProps} attributions={attributions} />);
		expect(
			wrap.containsMatchingElement(<meta name="citation_author" content="Weiwei Hsu" />),
		).toBeFalsy();
		expect(
			wrap.containsMatchingElement(<meta name="citation_author" content="Hugh Dubberly" />),
		).toBeTruthy();
		expect(
			wrap.containsMatchingElement(<meta name="citation_author" content="Ian Reynolds" />),
		).toBeTruthy();
	});

	it('generates citation_editor metadata for every contributor WITH a contributor role as their primary role', () => {
		const attributions = pubData.attributions.map((att) => ({ ...att }));
		attributions[1].roles = ['Series Editor', 'Test role'];
		attributions[2].roles = ['Test role', 'Writing â€“ Review & Editing'];

		const wrap = shallow(<MetaComponents {...defaultProps} attributions={attributions} />);
		expect(
			wrap.containsMatchingElement(<meta name="citation_editor" content="Weiwei Hsu" />),
		).toBeFalsy();
		expect(
			wrap.containsMatchingElement(<meta name="citation_editor" content="Hugh Dubberly" />),
		).toBeTruthy();
		expect(
			wrap.containsMatchingElement(<meta name="citation_editor" content="Ian Reynolds" />),
		).toBeFalsy();
	});

	it('generates the expected meta elements', () => {
		const props = {
			...defaultProps,
			attributions: [...attributionsData, ...pubData.attributions],
		};

		expect(generateMetaComponents(props as any)).toMatchSnapshot();
	});
});
