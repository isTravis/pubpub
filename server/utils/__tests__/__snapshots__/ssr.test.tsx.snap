// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMetaComponents generates the expected meta elements 1`] = `
Array [
  <link
    href="https://localhost/rss.xml"
    rel="alternate"
    title="Test Pub RSS Feed"
    type="application/rss+xml"
  />,
  <title>
    Test Pub
  </title>,
  <meta
    content="Test Pub"
    property="og:title"
  />,
  <meta
    content="Test Pub"
    name="twitter:title"
  />,
  <meta
    content="Test Pub"
    name="twitter:image:alt"
  />,
  <meta
    content="Test Pub"
    name="citation_title"
  />,
  <meta
    content="Test Pub"
    name="dc.title"
  />,
  <meta
    content="Joi Ito's PubPub"
    property="og:site_name"
  />,
  <meta
    content="Joi Ito's PubPub"
    name="citation_journal_title"
  />,
  <meta
    content="https://localhost/login"
    property="og:url"
  />,
  <meta
    content="website"
    property="og:type"
  />,
  <link
    href="https://assets.pubpub.org/12e422ul/41507920476077.png"
    rel="icon"
    sizes="256x256"
    type="image/png"
  />,
  Array [
    <meta
      content="Ian Reynolds"
      name="citation_author"
    />,
    <meta
      content="Travis Rich"
      name="citation_author"
    />,
    <meta
      content="Catherine Ahearn"
      name="citation_author"
    />,
    <meta
      content="Gabriel Stein"
      name="citation_author"
    />,
    <meta
      content="Joel Gustafson"
      name="citation_author"
    />,
    <meta
      content="Weiwei Hsu"
      name="citation_author"
    />,
    <meta
      content="Hugh Dubberly"
      name="citation_author"
    />,
    <meta
      content="Ian Reynolds"
      name="citation_author"
    />,
  ],
  Array [
    <meta
      content="Ian Reynolds"
      name="dc.creator"
    />,
    <meta
      content="Travis Rich"
      name="dc.creator"
    />,
    <meta
      content="Catherine Ahearn"
      name="dc.creator"
    />,
    <meta
      content="Gabriel Stein"
      name="dc.creator"
    />,
    <meta
      content="Joel Gustafson"
      name="dc.creator"
    />,
    <meta
      content="Weiwei Hsu"
      name="dc.creator"
    />,
    <meta
      content="Hugh Dubberly"
      name="dc.creator"
    />,
    <meta
      content="Ian Reynolds"
      name="dc.creator"
    />,
  ],
  Array [],
  <meta
    content="924988584221879"
    property="fb:app_id"
  />,
  <meta
    content="summary"
    name="twitter:card"
  />,
  <meta
    content="@pubpub"
    name="twitter:site"
  />,
]
`;
